pkg/processor/sourceprocessor/source_processor.go

```diff
@@ -162,151 +159,157 @@ func newSourceTraceProcessor(next consumer.Traces, cfg *Config) *sourceTraceProc
-
 		ilss := rs.InstrumentationLibrarySpans()
 		totalSpans := 0
 		for j := 0; j < ilss.Len(); j++ {
 			ils := ilss.At(j)
-			totalSpans = ils.Spans().Len()
+			totalSpans += ils.Spans().Len()
 		}
 
-		if stp.isFilteredOut(atts) {
+		if sp.isFilteredOut(atts) {
 			rs.InstrumentationLibrarySpans().RemoveIf(func(pdata.InstrumentationLibrarySpans) bool { return true })
 			observability.RecordFilteredOutN(totalSpans)
 		} else {
 			observability.RecordFilteredInN(totalSpans)
 		}
+	}
+
+	return td, nil
+}
+
+// ProcessMetrics processes metrics
+func (sp *sourceProcessor) ProcessMetrics(ctx context.Context, md pdata.Metrics) (pdata.Metrics, error) {
+	rss := md.ResourceMetrics()
+
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+		res := sp.processResource(rs.Resource())
+		atts := res.Attributes()
+
+		if sp.isFilteredOut(atts) {
+			rs.InstrumentationLibraryMetrics().RemoveIf(func(pdata.InstrumentationLibraryMetrics) bool { return true })
+		}
+	}
 
-		// Perhaps this is coming through Zipkin and in such case the attributes are stored in each span attributes, doh!
-		if !filledAnySource {
-			ilss := rs.InstrumentationLibrarySpans()
-			for j := 0; j < ilss.Len(); j++ {
-				ils := ilss.At(j)
-				inputSpans := ils.Spans()
-				outputSpans := pdata.NewSpanSlice()
-
-				for k := 0; k < inputSpans.Len(); k++ {
-					s := inputSpans.At(k)
-					atts := s.Attributes()
-
-					// TODO: move this to k8sprocessor
-					stp.enrichPodName(&atts)
-					if !filledOtherMeta {
-						stp.fillOtherMeta(atts)
-					}
-
-					stp.sourceHostFiller.fillResourceOrUseAnnotation(&atts, stp.annotationAttribute(sourceHostSpecialAnnotation), stp.keys)
-					stp.sourceCategoryFiller.fillResourceOrUseAnnotation(&atts, stp.annotationAttribute(sourceCategorySpecialAnnotation), stp.keys)
-					stp.sourceNameFiller.fillResourceOrUseAnnotation(&atts, stp.annotationAttribute(sourceNameSpecialAnnotation), stp.keys)
-
-					if !stp.isFilteredOut(atts) {
-						s.CopyTo(outputSpans.AppendEmpty())
-						observability.RecordFilteredIn()
-					} else {
-						observability.RecordFilteredOut()
-					}
-				}
-
-				ils.Spans().RemoveIf(func(pdata.Span) bool { return true })
-				outputSpans.MoveAndAppendTo(ils.Spans())
-			}
+	return md, nil
+}
+
+// ProcessLogs processes logs
+func (sp *sourceProcessor) ProcessLogs(ctx context.Context, md pdata.Logs) (pdata.Logs, error) {
+	rss := md.ResourceLogs()
+
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+		res := sp.processResource(rs.Resource())
+		atts := res.Attributes()
+
+		if sp.isFilteredOut(atts) {
+			rs.InstrumentationLibraryLogs().RemoveIf(func(pdata.InstrumentationLibraryLogs) bool { return true })
 		}
 	}
-	return stp.nextConsumer.ConsumeTraces(ctx, td)
+
+	return md, nil
+}
+
+// processResource perform multiple actions on resource:
+//   - enrich pod name, so it can be used in templates
+//   - fills source attributes based on config or annotations
+//   - set metadata (collector name)
```

Comment:
```suggestion
// processResource performs multiple actions on resource:
//   - enrich pod name, so it can be used in templates
//   - fills source attributes based on config or annotations
//   - set metadata (collector name)
```